File: ffi/secretary-ffi-uniffi/tests/kotlin/ConformanceErrors.kt
// Error variant name + detail extractors for VaultException.
//
// Enumerates every variant of VaultException. If uniffi adds a new variant,
// a new `is VaultException.XYZ` branch is needed here — the exhaustive
// `when` will fail to compile, acting as a tripwire the same way the
// Swift non-exhaustive-switch error does in ConformanceErrors.swift.

import uniffi.secretary.VaultException

internal fun vaultExceptionVariantName(e: VaultException): String = when (e) {
    is VaultException.WrongPasswordOrCorrupt -> "WrongPasswordOrCorrupt"
    is VaultException.WrongMnemonicOrCorrupt -> "WrongMnemonicOrCorrupt"
    is VaultException.InvalidMnemonic -> "InvalidMnemonic"
    is VaultException.VaultMismatch -> "VaultMismatch"
    is VaultException.CorruptVault -> "CorruptVault"
    is VaultException.FolderInvalid -> "FolderInvalid"
    is VaultException.BlockNotFound -> "BlockNotFound"
    is VaultException.InvalidArgument -> "InvalidArgument"
    is VaultException.SaveCryptoFailure -> "SaveCryptoFailure"
    is VaultException.NotAuthor -> "NotAuthor"
    is VaultException.RecipientAlreadyPresent -> "RecipientAlreadyPresent"
    is VaultException.RecipientNotPresent -> "RecipientNotPresent"
    is VaultException.CannotRevokeOwner -> "CannotRevokeOwner"
    is VaultException.MissingRecipientCard -> "MissingRecipientCard"
    is VaultException.CardDecodeFailure -> "CardDecodeFailure"
    is VaultException.BlockUuidAlreadyLive -> "BlockUuidAlreadyLive"
    is VaultException.BlockNotInTrash -> "BlockNotInTrash"
    is VaultException.BlockPurged -> "BlockPurged"
    is VaultException.RecordNotFound -> "RecordNotFound"
    is VaultException.ContactAlreadyExists -> "ContactAlreadyExists"
    is VaultException.ContactNotFound -> "ContactNotFound"
    is VaultException.CannotDeleteOwnerContact -> "CannotDeleteOwnerContact"
    is VaultException.SyncStateVaultMismatch -> "SyncStateVaultMismatch"
    is VaultException.SyncStateCorrupt -> "SyncStateCorrupt"
    is VaultException.SyncEvidenceStale -> "SyncEvidenceStale"
    is VaultException.SyncInProgress -> "SyncInProgress"
    is VaultException.SyncFailed -> "SyncFailed"
    is VaultException.SyncDecisionsIncomplete -> "SyncDecisionsIncomplete"
    is VaultException.DeviceSlotNotFound -> "DeviceSlotNotFound"
    is VaultException.WrongDeviceSecretOrCorrupt -> "WrongDeviceSecretOrCorrupt"
    is VaultException.DeviceUuidMismatch -> "DeviceUuidMismatch"
    is VaultException.VaultFolderNotEmpty -> "VaultFolderNotEmpty"
    is VaultException.VaultNeedsRepair -> "VaultNeedsRepair"
    is VaultException.RepairRejected -> "RepairRejected"
}

// Extract the detail string from VaultException variants that carry one.
// Returns null for variants that carry no detail (e.g. WrongPasswordOrCorrupt).
internal fun vaultExceptionDetail(e: VaultException): String? = when (e) {
    is VaultException.InvalidMnemonic -> e.detail
    is VaultException.CorruptVault -> e.detail
    is VaultException.FolderInvalid -> e.detail
    is VaultException.InvalidArgument -> e.detail
    is VaultException.SaveCryptoFailure -> e.detail
    is VaultException.CardDecodeFailure -> e.detail
    is VaultException.BlockUuidAlreadyLive -> e.detail
    is VaultException.BlockNotInTrash -> e.detail
    is VaultException.BlockPurged -> e.detail
    is VaultException.SyncStateCorrupt -> e.detail
    is VaultException.SyncFailed -> e.detail
    is VaultException.DeviceUuidMismatch -> e.detail
    is VaultException.RepairRejected -> e.detail
    // The remaining variants carry no detail string.
    is VaultException.WrongPasswordOrCorrupt,
    is VaultException.WrongMnemonicOrCorrupt,
    is VaultException.VaultMismatch,
    is VaultException.RecipientAlreadyPresent,
    is VaultException.RecipientNotPresent,
    is VaultException.CannotRevokeOwner,
    is VaultException.NotAuthor,
    is VaultException.MissingRecipientCard,
    is VaultException.BlockNotFound,
    is VaultException.RecordNotFound,
    // ContactAlreadyExists / ContactNotFound carry uuid_hex, not detail.
    is VaultException.ContactAlreadyExists,
    is VaultException.ContactNotFound,
    is VaultException.CannotDeleteOwnerContact,
    is VaultException.SyncStateVaultMismatch,
    is VaultException.SyncEvidenceStale,
    is VaultException.SyncInProgress,
    is VaultException.SyncDecisionsIncomplete,
    is VaultException.DeviceSlotNotFound,
    is VaultException.WrongDeviceSecretOrCorrupt,
    is VaultException.VaultFolderNotEmpty,
    // VaultNeedsRepair carries block_uuid_hex, not detail.
    is VaultException.VaultNeedsRepair -> null
}

File: ffi/secretary-ffi-uniffi/tests/swift/ConformanceErrors.swift
// Error variant name + detail extractors for VaultError.
//
// The match enumerates every variant of VaultError. If uniffi adds a new case
// the Swift compiler will emit a non-exhaustive-switch error — that's the
// intended tripwire.
//
// Mirrors the Rust variant_name_vault helper in core/tests/conformance_kat_helpers/errors.rs
// and the Kotlin vaultExceptionVariantName in ConformanceErrors.kt.

func vaultErrorName(_ e: VaultError) -> String {
    switch e {
    case .WrongPasswordOrCorrupt: return "WrongPasswordOrCorrupt"
    case .WrongMnemonicOrCorrupt: return "WrongMnemonicOrCorrupt"
    case .InvalidMnemonic: return "InvalidMnemonic"
    case .VaultMismatch: return "VaultMismatch"
    case .CorruptVault: return "CorruptVault"
    case .FolderInvalid: return "FolderInvalid"
    case .BlockNotFound: return "BlockNotFound"
    case .SaveCryptoFailure: return "SaveCryptoFailure"
    case .NotAuthor: return "NotAuthor"
    case .RecipientAlreadyPresent: return "RecipientAlreadyPresent"
    case .MissingRecipientCard: return "MissingRecipientCard"
    case .CardDecodeFailure: return "CardDecodeFailure"
    case .BlockUuidAlreadyLive: return "BlockUuidAlreadyLive"
    case .BlockNotInTrash: return "BlockNotInTrash"
    case .InvalidArgument: return "InvalidArgument"
    case .RecordNotFound: return "RecordNotFound"
    case .ContactAlreadyExists: return "ContactAlreadyExists"
    case .ContactNotFound: return "ContactNotFound"
    case .CannotDeleteOwnerContact: return "CannotDeleteOwnerContact"
    }
}

func vaultErrorDetail(_ e: VaultError) -> String? {
    switch e {
    case .InvalidMnemonic(let d): return d
    case .CorruptVault(let d): return d
    case .FolderInvalid(let d): return d
    case .SaveCryptoFailure(let d): return d
    case .CardDecodeFailure(let d): return d
    case .BlockUuidAlreadyLive(let d): return d
    case .BlockNotInTrash(let d): return d
    case .InvalidArgument(let d): return d
    default: return nil
    }
}
